airbyte-integrations/connectors/source-jira/source_jira/client.py

```diff
@@ -0,0 +1,84 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from functools import partial
+from json.decoder import JSONDecodeError
+from typing import Mapping, Tuple
+
+import requests
+from base_python import BaseClient
+from requests.auth import HTTPBasicAuth
+from requests.exceptions import ConnectionError
+
+
+class Client(BaseClient):
+    API_VERSION = 3
+    DEFAULT_ITEMS_PER_PAGE = 100
+
+    PARAMS = {"maxResults": DEFAULT_ITEMS_PER_PAGE, "startAt": 0}
+    ENTITIES_MAP = {
+        "projects": {"url": "/project/search", "func": lambda v: v["values"], "params": PARAMS},
+        "issues": {"url": "/search", "func": lambda v: v["issues"], "params": PARAMS},
+        "issue_comments": {"url": "/search", "func": lambda v: v["issues"], "params": {**PARAMS, **{"fields": ["comment"]}}},
+        "users": {"url": "/users/search", "func": lambda v: v, "params": PARAMS},
+        "resolutions": {"url": "/resolution", "func": lambda v: v, "params": {}},
+    }
+
+    def __init__(self, api_token, domain, email):
+        self.auth = HTTPBasicAuth(email, api_token)
+        self.base_api_url = f"https://{domain}/rest/api/{self.API_VERSION}"
+        super().__init__()
+
+    def lists(self, name, **kwargs):
+        while True:
+            response = requests.get(f"{self.base_api_url}{kwargs['url']}", params=kwargs["params"], auth=self.auth)
+            data = kwargs["func"](response.json())
+            if name == "issue_comments":
+                for obj in data:
+                    yield from obj["fields"]["comment"]["comments"]
+            else:
+                yield from data
+            if name == "resolutions" or len(data) < self.DEFAULT_ITEMS_PER_PAGE:
+                break
+            kwargs["params"]["startAt"] += self.DEFAULT_ITEMS_PER_PAGE
+
+    def _enumerate_methods(self) -> Mapping[str, callable]:
+        return {entity: partial(self.lists, name=entity, **value) for entity, value in self.ENTITIES_MAP.items()}
+
+    def health_check(self) -> Tuple[bool, str]:
+        alive = True
+        error_msg = None
+
+        try:
+            next(self.lists(name="resolutions", **self.ENTITIES_MAP["resolutions"]))
+
+        except ConnectionError as error:
+            alive, error_msg = False, str(error)
+        except JSONDecodeError:
```

Comment:
Why would there be a JSONDecodeError?

Reply:
This is only for check connect, because if somebody specify the wrong domain, then we will still receive a 200 status code in the response, but in the response.text we will not receive data, but text. We use the method of getting resolutions data to check the correct connection and getting data in the response, and if we do not receive the data in this case, then the incorrect credentials are entered.

Reply:
Got it. Added a suggested comment.